### ISSUE TITLE
Python API global variable prevents systematic creation of multiple models

### ISSUE THREAD

**Issue by paulromano** [mit-crpg/openmc#465]
When a universe is exported to XML, each cell in the universe is checked against the global variable WRITTEN_IDS to make sure it's not written twice. That's great, except that if you try to create multiple models in a single Python script, any cells with the same ID in the second or further models won't be written since they appear in WRITTEN_IDS. To get it to work correctly requires a hack:

``` python
import openmc.universe

# Write first model
...
openmc.universe.WRITTEN_IDS = {}

# Write another model
...
```

**Comment by wbinventor** [comment]
Yes, this is a known issue, one which is difficult to find a good solution for, but which would be nice to fix. Interestingly, @nhorelik and I discussed this very same problem as OpenCG's issue [#25](https://github.com/mit-crpg/OpenCG/issues/25) last year. As you can read in the discussion there, the approach I came up with to resolve this for OpenCG is the following. 

When initialized, each ``Material``, ``Cell``, ``Surface``, ``Universe``, ``Lattice``, etc. has a unique string UUID generated from the Python built-in ``uuid.uuid4()``. At this stage, none of these objects yet has an integer ID, unless one was directly specified for an object at initialization (but this sort of behavior is really no longer necessary). The user will then proceed to assemble the CSG tree culminating in a ``Geometry`` object as the root node. The OpenCG ``Geometry`` has a class method [``assignAutoIds()``](https://github.com/mit-crpg/OpenCG/blob/master/opencg/geometry.py#L276) which then assigns integer IDs to each of the objects contained within it. Effectively, the ``Geometry`` has separate "local ID streams" for ``Materials``, ``Surfaces``, etc. This is in place of "global ID streams" for each of these classes as we currently have in the OpenMC Python API. The local ID streams means that OpenCG no longer (or at least it shouldn't!) check for redundant IDs at the global level. 

If this doesn't make sense to you, or if you're confused by the algorithm, I highly suggest using an interactive debugger in an IDE like Eclipse or PyCharm to follow the flow of control for the auto local ID assignments in OpenCG. There are certainly some subtle complications that may not be easy to see directly from the code, but which should be self evident from a debugger.

I didn't implement this local ID stream concept into the OpenMC Python API since I wasn't confident at the time that it was an optimal or elegant solution to the problem. But at least it might be a starting point to solve this issue now.

**Comment by wbinventor** [comment]
I should mention that one subtle complication of the preceding algorithm is that the integer ID assignment is not reproducible. In particular, ``uuid.uuid4()`` will not generate the same UUIDs for each object from run-to-run - nor will they be "ordered" in the same way. In other words, if you generate 5 UUID strings one run 1, and 5 UUID strings on run 2, Python will not order them in the same way. The reason that this can be problematic (depending on your viewpoint) is that the OpenCG ``Geometry`` retrieves a ``dict`` of all ``Materials``, a ``dict`` of all ``Surfaces``, etc. in the CSG tree, with UUIDs as keys and object pointers as values. It then iterates over the dictionaries and assigns each integer IDs for each class type. Since the UUIDs are not strictly ordered from run-to-run, the integer IDs will be assigned to the objects in a different order each time one runs the same code.

In my mind, this is not an issue for either OpenCG or OpenMOC since integer IDs really have no relevance anyway (albeit, the are needed for some of the specialty algorithms in OpenCG). The integer IDs are only relevant for a code like OpenMC where a CSG tree is input implicitly through IDs, rather than explicitly through pointers connecting objects in the tree. Hence, it would probably be nice to have reproducible integer IDs for OpenMC since it can be tough debugging geometries which have different ID assignments from run-to-run.

**Comment by paulromano** [comment]
Hmm, ok. Thanks for elaborating on what was done for OpenCG. I'll have to cogitate on this to decide if we should do the same for OpenMC.

"This is a known issue" --> only known once someone submits an issue! :)

**Comment by paulromano** [comment]
I believe this can be closed since we got rid of WRITTEN_IDS in 357aa894c8a65fcb595cbe6ac382dee0413ccfb7.
